src/bottom_topography/spline_cache_2D.jl

```diff
@@ -0,0 +1,360 @@
+      idx = 0
+      for i in (inner_elmts+(2*m)+1):(inner_elmts+(2*m)+n)
+        Phi[i, idx           + 1] = - 1
+        Phi[i, idx           + 2] = - 4
+        Phi[i, idx           + 3] = - 1
+        Phi[i, idx +   (n+2) + 1] =   4
+        Phi[i, idx +   (n+2) + 2] =  16
+        Phi[i, idx +   (n+2) + 3] =   4
+        Phi[i, idx + 2*(n+2) + 1] = - 6
+        Phi[i, idx + 2*(n+2) + 2] = -24
+        Phi[i, idx + 2*(n+2) + 3] = - 6
+        Phi[i, idx + 3*(n+2) + 1] =   4
+        Phi[i, idx + 3*(n+2) + 2] =  16
+        Phi[i, idx + 3*(n+2) + 3] =   4
+        Phi[i, idx + 4*(n+2) + 1] = - 1
+        Phi[i, idx + 4*(n+2) + 2] = - 4
+        Phi[i, idx + 4*(n+2) + 3] = - 1
+        idx += 1
+      end
+
+      # lower edge (u+)
+      idx = (m-3) * (n+2)
+      for i in (inner_elmts+(2*m)+(n+1)):(inner_elmts+(2*m)+(2*n))
+        Phi[i, idx           + 1] = - 1
+        Phi[i, idx           + 2] = - 4
+        Phi[i, idx           + 3] = - 1
+        Phi[i, idx +   (n+2) + 1] =   4
+        Phi[i, idx +   (n+2) + 2] =  16
+        Phi[i, idx +   (n+2) + 3] =   4
+        Phi[i, idx + 2*(n+2) + 1] = - 6
+        Phi[i, idx + 2*(n+2) + 2] = -24
+        Phi[i, idx + 2*(n+2) + 3] = - 6
+        Phi[i, idx + 3*(n+2) + 1] =   4
+        Phi[i, idx + 3*(n+2) + 2] =  16
+        Phi[i, idx + 3*(n+2) + 3] =   4
+        Phi[i, idx + 4*(n+2) + 1] = - 1
+        Phi[i, idx + 4*(n+2) + 2] = - 4
+        Phi[i, idx + 4*(n+2) + 3] = - 1
+        idx += 1
+      end
+
+      ## corners ##
+      i = inner_elmts + boundary_elmts - 3   
+    
+      # upper left corner
+      Phi[(i  ),               1] =  1
+      Phi[(i  ),               2] = -1
+      Phi[(i  ),       (n+2) + 1] = -1
+      Phi[(i  ),       (n+2) + 2] =  1
+      # upper right corner
+      Phi[(i+1),       (n+2) - 1] = -1
+      Phi[(i+1),       (n+2)    ] =  1
+      Phi[(i+1),     2*(n+2) - 1] =  1
+      Phi[(i+1),     2*(n+2)    ] = -1
+      # lower left corner
+      Phi[(i+2), (m  )*(n+2) + 1] = -1
+      Phi[(i+2), (m  )*(n+2) + 2] =  1
+      Phi[(i+2), (m+1)*(n+2) + 1] =  1
+      Phi[(i+2), (m+1)*(n+2) + 2] = -1
+      # lower right corner
+      Phi[(i+3), (m+1)*(n+2) - 1] =  1
+      Phi[(i+3), (m+1)*(n+2)    ] = -1
+      Phi[(i+3), (m+2)*(n+2) - 1] = -1
+      Phi[(i+3), (m+2)*(n+2)    ] =  1
+
+      Q_temp = 36 * (Phi\P)
+      Q      = reshape(Q_temp, (n+2, m+2)) 
+
+      BicubicBSpline(x, y, z, h, Q, IP)
+    end
+    
+  else
+    @error("Only free and not-a-knot boundary conditions are implemented!")
+  
+  end
+
+end
+
+# Read from file
+function bicubic_b_spline(path::String; boundary = "free", smoothing_factor = 0.0)
+  file = open(path)
+  lines = readlines(file)
+  close(file)
+
+  n = parse(Int64, lines[2])
+  m = parse(Int64, lines[4])
+  
+  x     = [parse(Float64, val) for val in lines[6      :5+n    ]]
+  y     = [parse(Float64, val) for val in lines[(7+n)  :(6+n+m)]]
+  z_tmp = [parse(Float64, val) for val in lines[(8+n+m):end    ]]
```

Comment:
similar to the comment for the 1D file readin. It should stated somewhere that you assumed a particular format to the way the spline data to be read in has been saved.

Reply:
Has been documented